pkg/clusterregistry/testing/testserver.go

```diff
@@ -0,0 +1,159 @@
+		}
+		if err != nil && !strings.Contains(err.Error(), "bind") {
+			break
+		}
+		t.Logf("Bind error, retrying...")
+	}
+
+	t.Fatalf("Failed to launch server")
+	return nil, nil
+}
+
+// startTestServer starts a etcd server and cluster registry API server. Returns
+// a REST client config and a tear-down function.
+//
+// Note: we return a tear-down func instead of a stop channel because the latter
+//       will leak temporary files that becaues Golang testing's call to os.Exit
+//       will not give a stop channel go routine enough time to remove
+//       temporary files.
+func startTestServer(t *testing.T) (result *restclient.Config, tearDownForCaller TearDownFunc, err error) {
+	var tmpDir string
+	var etcdServer *etcdtesting.EtcdTestServer
+	stopCh := make(chan struct{})
+	tearDown := func() {
+		close(stopCh)
+		if etcdServer != nil {
+			etcdServer.Terminate(t)
+		}
+		if len(tmpDir) != 0 {
+			os.RemoveAll(tmpDir)
+		}
+	}
+	defer func() {
+		if tearDownForCaller == nil {
+			tearDown()
+		}
+	}()
+
+	t.Logf("Starting etcd...")
+	etcdServer, storageConfig := etcdtesting.NewUnsecuredEtcd3TestClientServer(t, install.Scheme)
+
+	tmpDir, err = ioutil.TempDir("", "cluster-registry-clusterregistry-apiserver")
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create temp dir: %v", err)
+	}
+	port, err := freePort()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to find a free port: %v", err)
+	}
+	s := options.NewServerRunOptions()
+	s.SecureServing.BindPort = port
+	s.SecureServing.ServerCert.CertDirectory = tmpDir
+	s.Etcd.StorageConfig = *storageConfig
+	s.Etcd.DefaultStorageMediaType = "application/json"
+
+	t.Logf("Starting cluster registry...")
+	server, err := clusterregistry.CreateServer(s)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create server: %v", err)
+	}
+	runErrCh := make(chan error, 1)
+	go func(stopCh <-chan struct{}) {
+		if err := server.PrepareRun().Run(stopCh); err != nil {
+			t.Logf("cluster registry exited uncleanly: %v", err)
+			runErrCh <- err
+		}
+	}(stopCh)
+
+	t.Logf("Waiting for /healthz to be OK...")
+	client, err := kubernetes.NewForConfig(server.LoopbackClientConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create a client: %v", err)
+	}
+	err = wait.Poll(100*time.Millisecond, 30*time.Second, func() (bool, error) {
+		select {
+		case err := <-runErrCh:
+			return false, err
+		default:
+		}
+
+		result := client.CoreV1().RESTClient().Get().AbsPath("/healthz").Do()
+		status := 0
+		result.StatusCode(&status)
+		if status == 200 {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to wait for /healthz to return OK: %v", err)
+	}
+
+	return server.LoopbackClientConfig, tearDown, nil
+}
+
+// freePort finds and returns a free port on the local machine. Will return -1
+// for the port if there is an error.
+func freePort() (int, error) {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
```

Comment:
This doesn't seem consistent with the comment in line 44 - https://github.com/kubernetes/cluster-registry/pull/34/files#diff-096c0acbe4241fc9010c8fd6e54c914bR44. Am I missing something?

Reply:
The comment in line 44 is talking about `options.SecureServing.BindPort`, whose semantics are different: for that field, if it is set to 0, then HTTPS serving is disabled rather than the system attempting to find a port. Setting the port to 0 as a way to find an open port is for the system libraries, which is why it's being done here.

Reply:
I see Ok. Didn't realize that.